File: 56-merge-intervals/merge-intervals.py
class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: List[List[int]]
        """
        intervals.sort(key = lambda l: l[0])
        res = []

        slow = 0
        fast = 0

        while fast < len(intervals):
            startInterval = intervals[slow][0]
            endInterval = intervals[slow][1]

            while fast < len(intervals) and intervals[fast][0] <= endInterval:
                if intervals[fast][1] >= endInterval:
                    endInterval = intervals[fast][1]
                    
                fast += 1

            res.append([startInterval, endInterval])
            slow = fast

        return res
        

File: 56-merge-intervals/merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        List<int[]> res = new ArrayList<>();

        int slow = 0;
        int fast = 0;

        while(fast < intervals.length) {
            int startInterval = intervals[slow][0];
            int endInterval = intervals[slow][1];

            while (fast < intervals.length && intervals[fast][0] <= endInterval) {
                if (intervals[fast][1] >= endInterval) {
                    endInterval = intervals[fast][1];
                }

                fast++;
            }

            res.add(new int[]{startInterval, endInterval});
            slow = fast;
        }

        
        return res.toArray(new int[res.size()][]);

    }
}